internal/helpers/poll.go

```diff
@@ -2,23 +2,163 @@ package helpers
 
 import (
 	"context"
+	"sync"
 	"time"
 
+	"k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/util/wait"
 )
 
+func PollUntilContextCancel(ctx context.Context, getInterval func() time.Duration, immediate bool, condition wait.ConditionWithContextFunc) error {
+	return loopConditionUntilContext(ctx, NewDynamicTimer(getInterval), immediate, false, condition)
+}
+
 // BackgroundPollUntilContextCancel spawns a new goroutine that runs the condition function on interval.
 // If syncFirstRun is set to true, it will execute the condition function synchronously first and then start
 // polling. Since error is returned synchronously, the only way to check for it is to use syncFirstRun.
 // Background poller does not sync errors. It can be stopped externally by cancelling the provided context.
-func BackgroundPollUntilContextCancel(ctx context.Context, interval time.Duration, immediate, syncFirstRun bool, condition wait.ConditionWithContextFunc) (err error) {
+func BackgroundPollUntilContextCancel(ctx context.Context, getInterval func() time.Duration, immediate, syncFirstRun bool, condition wait.ConditionWithContextFunc) (err error) {
 	if syncFirstRun {
 		_, err = condition(ctx)
 	}
 
 	go func() {
-		_ = wait.PollUntilContextCancel(ctx, interval, immediate, condition)
+		timer := NewDynamicTimer(getInterval)
+		_ = loopConditionUntilContext(ctx, timer, immediate, false, condition)
 	}()
 
 	return err
 }
+
+// loopConditionUntilContext executes the provided condition at intervals defined by
+// the provided timer until the provided context is cancelled, the condition returns
+// true, or the condition returns an error. If sliding is true, the period is computed
+// after condition runs. If it is false then period includes the runtime for condition.
+// If immediate is false the first delay happens before any call to condition, if
+// immediate is true the condition will be invoked before waiting and guarantees that
+// the condition is invoked at least once, regardless of whether the context has been
+// cancelled. The returned error is the error returned by the last condition or the
+// context error if the context was terminated.
+//
+// This is the common loop construct for all polling in the wait package.
+func loopConditionUntilContext(ctx context.Context, t wait.Timer, immediate, sliding bool, condition wait.ConditionWithContextFunc) error {
```

Comment:
should we get rid of the sliding param?  It looks like it's not used and somewhat confusing

Reply:
done